### PR TITLE
Testing cb_login route

### PR DESCRIPTION
**Pull request by astroash**
<span class="emoji-outer emoji-sizer"><span class="emoji-inner" style="background: url(chrome-extension://immhpnclomdloikkpcefncmfgjbkojmh/emoji-data/sheet_apple_32.png);background-position:28.025851938895418% 81.96239717978848%;background-size:5418.75% 5418.75%" data-codepoints="1f44b"></span></span> LOOK MOM - I wrote a test <span class="emoji-outer emoji-sizer"><span class="emoji-inner" style="background: url(chrome-extension://immhpnclomdloikkpcefncmfgjbkojmh/emoji-data/sheet_apple_32.png);background-position:28.025851938895418% 81.96239717978848%;background-size:5418.75% 5418.75%" data-codepoints="1f44b"></span></span>
- Add test user to db
- Return 401 if login is unsuccessful
- Test both successful and unsuccessful login requests
- Copy and marvel at some other sneaky test stuffs
     * @finnhodgkin's ~~super smart~~ ~~attention seeking~~ still-in-the-codebase hidden process killer
     * @finnhodgkin's ~~beautiful~~ immature lack of naming

### CODE REVIEW

File: tests/routes_tests/cbauthentication_tests/cb_login.test.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+const dbConnection = require('../../../react-backend/database/dbConnection.js');
+const test = require('tape');
+const request = require('supertest');
+const app = require('../../../react-backend/app');
+
+test('Route cblogin | post | successful request', (t) => {
```
Comment (eliascodes) [inline tests/routes_tests/cbauthentication_tests/cb_login.test.js]: When doing API testing I typically put the URL first here, for example:
```js
test('POST /api/cb/login | successful request', (t) => {});
```

What does "successful request" mean here in terms of the scenario we are testing? Is this a request where the authentication details are accepted? If so, I'd suggest being more specific, e.g. "authentication successful".

File: tests/routes_tests/cbauthentication_tests/cb_login.test.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  t.plan(3);
+  const successPayload = { formEmail: 'findmyfroggy@frogfinders.com', formPswd: 'Funnyfingers11!' };
+  const failPayload = { formEmail: 'findmyfroggy@frogfinders.com', formPswd: 'password123' };
```
Comment (eliascodes) [inline tests/routes_tests/cbauthentication_tests/cb_login.test.js]: If this test case is about a "successful request", why are we also posting payloads that return nominally unsuccessful responses?

File: tests/routes_tests/cbauthentication_tests/cb_login.test.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    .send(successPayload)
+    .expect(200)
+    .expect('Content-Type', /json/)
+    .end((err, res) => {
+      t.ok(res.body.success, 'cb successfully logged in with correct password');
+    });
+  request(app)
+    .post('/api/cb/login')
+    .send(failPayload)
+    .expect(200)
```
Comment (eliascodes) [inline tests/routes_tests/cbauthentication_tests/cb_login.test.js]: It doesn't seem right that a request where authentication is rejected returns a `200`. `401` is the code I would expect. Why is this the case?

Comment (astroash) [reply tests/routes_tests/cbauthentication_tests/cb_login.test.js]: Looks like the server loves dishing out `200`'s. I'll change files individually as I write route tests.

Comment (eliascodes) [reply tests/routes_tests/cbauthentication_tests/cb_login.test.js]: Be careful, the client app may rely on the fact that 200 codes are being returned 😕

File: tests/routes_tests/cbauthentication_tests/cb_login.test.js (hunk 1 above)
Comment (astroash) [reply tests/routes_tests/cbauthentication_tests/cb_login.test.js]: Hmm... I was vaguely going along the lines of it being tests for when a user makes a successful POST request to this route, but I guess that makes less sense when I type it out as that's what supertest is mocking. Happy for suggestions on what you think it should be called.

File: tests/routes_tests/cbauthentication_tests/cb_login.test.js (hunk 2 above)
Comment (astroash) [reply tests/routes_tests/cbauthentication_tests/cb_login.test.js]: ☝️ see my previous waffle...

File: tests/routes_tests/cbauthentication_tests/cb_login.test.js (hunk 1 above)
Comment (eliascodes) [reply tests/routes_tests/cbauthentication_tests/cb_login.test.js]: > "authentication successful"

☝️ is fine, with the test _only_ testing a request with a payload that results in successful authentication.

Other scenarios should be handled in separate test cases.

File: react-backend/routes/cbauthentication/cb_register.js [hunk 4]
```diff
@@ -7,7 +7,10 @@ router.post('/', (req, res, next) => {
   const { formPswd, formName, formEmail, formGenre } = req.body;
 
   const hashedPassword = hashCB(formPswd);
+  console.log('pw', hashedPassword);
```
Comment (eliascodes) [inline react-backend/routes/cbauthentication/cb_register.js]: do we need these logs?

File: react-backend/routes/cbauthentication/cb_login.js [hunk 5]
```diff
@@ -20,10 +19,14 @@ router.post('/', (req, res, next) => {
...
-      const loggedIn = exists ? { success: true, token } : { success: false };
-
-      res.send(loggedIn);
+      if (exists) {
+        const token = jwt.sign({ email: formEmail }, process.env.SECRET);
+        const loggedInResponse = { success: true, token };
+        res.send(loggedInResponse);
+      } else {
+        const errorResponse = { success: false };
+        res.status(401).send(errorResponse);
```
Comment (eliascodes) [inline react-backend/routes/cbauthentication/cb_login.js]: As far as I can see, no part of the client code explicitly relies on this route returning a `200` code for an unsuccessful response, but could you confirm you've double checked manually (e.g. by clicking through and trying yourself?)

Comment (astroash) [reply react-backend/routes/cbauthentication/cb_login.js]: Totes did, I thought I had done the bad thing, but turns out it was already broked. See #207

Comment (eliascodes) [reply react-backend/routes/cbauthentication/cb_login.js]: Cool, thanks, just the logs then

File: react-backend/routes/cbauthentication/cb_register.js (hunk 4 above)
Comment (astroash) [reply react-backend/routes/cbauthentication/cb_register.js]: Lol, that doesn't need to be a question.

**Comment by eliascodes** [comment]
@astroash there's nothing for me to do on this. I've approved. Second reviewer to merge, which is @RogeredBacon

**Comment by astroash** [comment]
@eliascodes 🙈 sorry, meant to do that in the other PR

File: package.json [hunk 6]
```diff
@@ -11,7 +11,7 @@
     "heroku-postbuild": "cd react-backend/client && npm install --only=dev && npm install && npm run build",
     "build.client": "cd react-backend/client && npm run build",
     "db_build": "node ./react-backend/database/dbBuild",
-    "test": "NODE_ENV=test tape tests/*.test.js | tap-spec",
+    "test": "NODE_ENV=test nyc tape tests/**/*.test.js | tap-spec",
```
Comment (eliascodes) [inline package.json]: Sorry, can you remove the `nyc` here please?

Comment (eliascodes) [reply package.json]: But also, you're not meant to be working today. 🤔

**Comment by RogeredBacon** [comment]
Travis is saying:

/home/travis/build/TwinePlatform/DataPower/tests/routes_tests/cbauthentication_tests/cb_login.test.js
  1:7  error  'dbConnection' is assigned a value but never used  no-unused-vars

Is this expected?

**Comment by eliascodes** [comment]
@RogeredBacon yes, the linter returned an error, so the build failed. The system works!

**Comment by RogeredBacon** [comment]
Success! I've removed the nyc, is this ready to merge then?

**Comment by eliascodes** [comment]
@RogeredBacon no, the build must be successful for this to be merged. That means the linter error needs to be fixed 😄

**Comment by RogeredBacon** [comment]
Oh sorry I thought you meant that failure was expected haha.

Ok I'll fix it

**Comment by astroash** [comment]
@RogeredBacon, I'm happy to do it if ya want.

**Comment by RogeredBacon** [comment]
Nah it's your day off and its an easy fix

**Pull request merged**